File: backend/app/services/opportunity_service.py
"""Opportunity service"""
from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, or_
from datetime import datetime
from decimal import Decimal

from app.models.opportunity import Opportunity, OpportunityContact
from app.models.contact import Contact
from app.models.activity import Activity
from app.models.document import Document


async def create_opportunity(
    db: AsyncSession,
    tenant_id: str,
    data: Dict[str, Any],
) -> Opportunity:
    """Create a new opportunity"""
    def _normalize_datetime(value: Any) -> Optional[datetime]:
        if value is None or value == "":
            return None
        if isinstance(value, datetime):
            if value.tzinfo is not None:
                from datetime import timezone
                return value.astimezone(timezone.utc).replace(tzinfo=None)
            return value
        if isinstance(value, str):
            try:
                parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
                if parsed.tzinfo is not None:
                    from datetime import timezone
                    parsed = parsed.astimezone(timezone.utc).replace(tzinfo=None)
                return parsed
            except (ValueError, AttributeError, TypeError):
                return None
        return None

    def _to_decimal(value: Any) -> Optional[Decimal]:
        if value is None or value == "":
            return None
        try:
            return Decimal(str(value))
        except (ValueError, TypeError):
            return None

    due_date = _normalize_datetime(data.get("due_date"))
    rfp_submission_date = _normalize_datetime(data.get("rfp_submission_date"))
    award_date = _normalize_datetime(data.get("award_date"))
    next_task_due = _normalize_datetime(data.get("next_task_due"))

    value = _to_decimal(data.get("value"))
    pwin = _to_decimal(data.get("pwin"))
    ptw = _to_decimal(data.get("ptw"))

    number_of_years = None
    if data.get("number_of_years") not in (None, ""):
        try:
            number_of_years = int(data["number_of_years"])
        except (ValueError, TypeError):
            number_of_years = None

    opportunity = Opportunity(
        tenant_id=tenant_id,
        name=data["name"],
        agency=data.get("agency"),
        sub_agency=data.get("sub_agency"),
        stage=data.get("stage", "qualification"),
        status=data.get("status", "active"),  # Explicitly set status
        value=value,
        pwin=pwin,
        ptw=ptw,
        due_date=due_date,
        rfp_submission_date=rfp_submission_date,
        award_date=award_date,
        naics_code=data.get("naics_code"),
        contract_vehicle=data.get("contract_vehicle"),
        opportunity_type=data.get("opportunity_type"),
        description=data.get("description"),
        summary=data.get("summary"),
        history_notes=data.get("history_notes"),
        next_task_comments=data.get("next_task_comments"),
        next_task_due=next_task_due,
        capture_manager=data.get("capture_manager"),
        agency_pocs=data.get("agency_pocs"),
        business_sectors=data.get("business_sectors"),
        role=data.get("role"),
        number_of_years=number_of_years,
        bd_status=data.get("bd_status"),
        requirements=data.get("requirements"),
        account_id=data.get("account_id"),
        owner_id=data.get("owner_id"),
    )
    
    db.add(opportunity)
    await db.commit()
    await db.refresh(opportunity)
    return opportunity


async def get_opportunity(
    db: AsyncSession,
    opportunity_id: str,
    tenant_id: str,
) -> Optional[Opportunity]:
    """Get an opportunity by ID"""
    result = await db.execute(
        select(Opportunity).where(
            and_(
                Opportunity.id == opportunity_id,
                Opportunity.tenant_id == tenant_id,
            )
        )
    )
    return result.scalar_one_or_none()


async def list_opportunities(
    db: AsyncSession,
    tenant_id: str,
    filters: Optional[Dict[str, Any]] = None,
    skip: int = 0,
    limit: int = 100,
) -> List[Opportunity]:
    """List opportunities with filtering"""
    query = select(Opportunity).where(Opportunity.tenant_id == tenant_id)
    
    if filters:
        if filters.get("stage"):
            query = query.where(Opportunity.stage == filters["stage"])
        if filters.get("status"):
            query = query.where(Opportunity.status == filters["status"])
        if filters.get("agency"):
            query = query.where(Opportunity.agency.ilike(f"%{filters['agency']}%"))
        if filters.get("owner_id"):
            query = query.where(Opportunity.owner_id == filters["owner_id"])
    
    query = query.offset(skip).limit(limit).order_by(Opportunity.created_at.desc())
    result = await db.execute(query)
    return list(result.scalars().all())


async def update_opportunity(
    db: AsyncSession,
    opportunity_id: str,
    tenant_id: str,
    data: Dict[str, Any],
) -> Optional[Opportunity]:
    """Update an opportunity"""
    opportunity = await get_opportunity(db, opportunity_id, tenant_id)
    if not opportunity:
        return None
    
    def _normalize_update_datetime(value: Any) -> Optional[datetime]:
        if value is None or value == "":
            return None
        if isinstance(value, datetime):
            if value.tzinfo is not None:
                from datetime import timezone
                return value.astimezone(timezone.utc).replace(tzinfo=None)
            return value
        if isinstance(value, str):
            try:
                parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
                if parsed.tzinfo is not None:
                    from datetime import timezone
                    parsed = parsed.astimezone(timezone.utc).replace(tzinfo=None)
                return parsed
            except (ValueError, AttributeError, TypeError):
                return None
        return None

    for key, value in data.items():
        if not hasattr(opportunity, key):
            continue

        if key in {"value", "pwin", "ptw"}:
            converted = _to_decimal(value)
            setattr(opportunity, key, converted)
        elif key in {"due_date", "rfp_submission_date", "award_date", "next_task_due"}:
            setattr(opportunity, key, _normalize_update_datetime(value))
        elif key == "number_of_years":
            if value in (None, ""):
                setattr(opportunity, key, None)
            else:
                try:
                    setattr(opportunity, key, int(value))
                except (ValueError, TypeError):
                    pass
        elif value is not None:
            setattr(opportunity, key, value)
 
    opportunity.updated_at = datetime.utcnow()
    await db.commit()
    await db.refresh(opportunity)
    return opportunity


async def add_contact_to_opportunity(
    db: AsyncSession,
    opportunity_id: str,
    contact_id: str,
    role: str,
    tenant_id: str,
) -> Optional[OpportunityContact]:
    """Add a contact to an opportunity"""
    # Verify opportunity belongs to tenant
    opportunity = await get_opportunity(db, opportunity_id, tenant_id)
    if not opportunity:
        return None
    
    opp_contact = OpportunityContact(
        opportunity_id=opportunity_id,
        contact_id=contact_id,
        role=role,
    )
    
    db.add(opp_contact)
    await db.commit()
    await db.refresh(opp_contact)
    return opp_contact


async def add_activity(
    db: AsyncSession,
    tenant_id: str,
    user_id: str,
    opportunity_id: Optional[str],
    activity_type: str,
    subject: Optional[str],
    description: Optional[str],
) -> Activity:
    """Add an activity to timeline"""
    activity = Activity(
        tenant_id=tenant_id,
        user_id=user_id,
        opportunity_id=opportunity_id,
        activity_type=activity_type,
        subject=subject,
        description=description,
        activity_date=datetime.utcnow(),
    )
    
    db.add(activity)
    await db.commit()
    await db.refresh(activity)
    return activity


async def get_opportunity_timeline(
    db: AsyncSession,
    opportunity_id: str,
    tenant_id: str,
) -> List[Activity]:
    """Get activity timeline for an opportunity"""
    result = await db.execute(
        select(Activity).where(
            and_(
                Activity.opportunity_id == opportunity_id,
                Activity.tenant_id == tenant_id,
            )
        ).order_by(Activity.activity_date.desc())
    )
    return list(result.scalars().all())

File: frontend/src/services/opportunityService.ts
import api from './api'
import { logApiCall, logApiResponse, logApiError } from '../utils/debug'

export interface Opportunity {
  id: string
  name: string
  agency?: string
  sub_agency?: string
  stage: string
  status: string
  value?: number
  pwin?: number
  ptw?: number
  due_date?: string
  rfp_submission_date?: string
  award_date?: string
  summary?: string
  history_notes?: string
  next_task_comments?: string
  next_task_due?: string
  capture_manager?: string
  agency_pocs?: string
  business_sectors?: string
  role?: string
  number_of_years?: number
  bd_status?: string
  naics_code?: string
  contract_vehicle?: string
  opportunity_type?: string
  description?: string
  requirements?: string
  account_id?: string | null
  owner_id?: string
  created_at: string
  updated_at: string
}

export const opportunityService = {
  list: async (params?: {
    skip?: number
    limit?: number
    stage?: string
    status?: string
    agency?: string
  }): Promise<{ opportunities: Opportunity[]; total: number }> => {
    const response = await api.get<{ opportunities: Opportunity[]; total: number }>(
      '/opportunities/opportunities',
      { params }
    )
    return response.data
  },

  get: async (id: string): Promise<Opportunity> => {
    const response = await api.get<Opportunity>(`/opportunities/opportunities/${id}`)
    return response.data
  },

  create: async (data: Partial<Opportunity>): Promise<Opportunity> => {
    // Convert data to match backend schema
    const payload: any = {
      name: data.name,
      stage: data.stage || 'qualification',
      status: data.status || 'active',
    }
    
    // Only include fields that are provided
    if (data.agency !== undefined && data.agency !== null && data.agency !== '') {
      payload.agency = data.agency
    }
    if (data.sub_agency) {
      payload.sub_agency = data.sub_agency
    }
    if (data.bd_status) {
      payload.bd_status = data.bd_status
    }
    if (data.summary) {
      payload.summary = data.summary
    }
    if (data.history_notes) {
      payload.history_notes = data.history_notes
    }
    if (data.next_task_comments) {
      payload.next_task_comments = data.next_task_comments
    }
    if (data.capture_manager) {
      payload.capture_manager = data.capture_manager
    }
    if (data.agency_pocs) {
      payload.agency_pocs = data.agency_pocs
    }
    if (data.business_sectors) {
      payload.business_sectors = data.business_sectors
    }
    if (data.role) {
      payload.role = data.role
    }
    if (data.contract_vehicle) {
      payload.contract_vehicle = data.contract_vehicle
    }
    if (data.naics_code) {
      payload.naics_code = data.naics_code
    }
    if (data.opportunity_type) {
      payload.opportunity_type = data.opportunity_type
    }
    if (data.description) {
      payload.description = data.description
    }
    if (data.requirements) {
      payload.requirements = data.requirements
    }
    if (data.account_id) {
      payload.account_id = data.account_id
    }
    if (data.owner_id) {
      payload.owner_id = data.owner_id
    }
    if (data.value !== undefined && data.value !== null) {
      payload.value = Number(data.value)
    }
    if (data.number_of_years !== undefined && data.number_of_years !== null) {
      payload.number_of_years = Number(data.number_of_years)
    }
    if (data.pwin !== undefined && data.pwin !== null) {
      payload.pwin = Number(data.pwin)
    }
    if (data.ptw !== undefined && data.ptw !== null) {
      payload.ptw = Number(data.ptw)
    }
    // due_date should be in ISO format if provided
    if (data.due_date) {
      const dateStr = data.due_date.includes('T') ? data.due_date : `${data.due_date}T00:00:00`
      payload.due_date = new Date(dateStr).toISOString()
    }
    if (data.rfp_submission_date) {
      const dateStr = data.rfp_submission_date.includes('T') ? data.rfp_submission_date : `${data.rfp_submission_date}T00:00:00`
      payload.rfp_submission_date = new Date(dateStr).toISOString()
    }
    if (data.next_task_due) {
      const dateStr = data.next_task_due.includes('T') ? data.next_task_due : `${data.next_task_due}T00:00:00`
      payload.next_task_due = new Date(dateStr).toISOString()
    }
    if (data.award_date) {
      const dateStr = data.award_date.includes('T') ? data.award_date : `${data.award_date}T00:00:00`
      payload.award_date = new Date(dateStr).toISOString()
    }
    
    logApiCall('POST', '/opportunities/opportunities', payload)
    
    try {
      const response = await api.post<Opportunity>('/opportunities/opportunities', payload)
      logApiResponse(response)
      return response.data
    } catch (error) {
      logApiError(error)
      throw error
    }
  },

  update: async (id: string, data: Partial<Opportunity>): Promise<Opportunity> => {
    // Convert data to match backend schema
    const payload: any = {}
    
    if (data.name !== undefined) payload.name = data.name
    if (data.agency !== undefined && data.agency !== null && data.agency !== '') {
      payload.agency = data.agency
    }
    if (data.sub_agency !== undefined) {
      payload.sub_agency = data.sub_agency
    }
    if (data.stage !== undefined) payload.stage = data.stage
    if (data.status !== undefined) payload.status = data.status
    if (data.bd_status !== undefined) payload.bd_status = data.bd_status
    if (data.summary !== undefined) payload.summary = data.summary
    if (data.history_notes !== undefined) payload.history_notes = data.history_notes
    if (data.next_task_comments !== undefined) payload.next_task_comments = data.next_task_comments
    if (data.capture_manager !== undefined) payload.capture_manager = data.capture_manager
    if (data.agency_pocs !== undefined) payload.agency_pocs = data.agency_pocs
    if (data.business_sectors !== undefined) payload.business_sectors = data.business_sectors
    if (data.role !== undefined) payload.role = data.role
    if (data.contract_vehicle !== undefined) payload.contract_vehicle = data.contract_vehicle
    if (data.naics_code !== undefined) payload.naics_code = data.naics_code
    if (data.opportunity_type !== undefined) payload.opportunity_type = data.opportunity_type
    if (data.description !== undefined) payload.description = data.description
    if (data.requirements !== undefined) payload.requirements = data.requirements
    if (data.account_id !== undefined) payload.account_id = data.account_id
    if (data.owner_id !== undefined) payload.owner_id = data.owner_id
    if (data.value !== undefined && data.value !== null) {
      payload.value = Number(data.value)
    }
    if (data.number_of_years !== undefined && data.number_of_years !== null) {
      payload.number_of_years = Number(data.number_of_years)
    }
    if (data.pwin !== undefined && data.pwin !== null) {
      payload.pwin = Number(data.pwin)
    }
    if (data.ptw !== undefined && data.ptw !== null) {
      payload.ptw = Number(data.ptw)
    }
    if (data.due_date !== undefined && data.due_date !== null && data.due_date !== '') {
      const dateStr = data.due_date.includes('T') ? data.due_date : `${data.due_date}T00:00:00`
      payload.due_date = new Date(dateStr).toISOString()
    }
    if (data.rfp_submission_date !== undefined && data.rfp_submission_date !== null && data.rfp_submission_date !== '') {
      const dateStr = data.rfp_submission_date.includes('T') ? data.rfp_submission_date : `${data.rfp_submission_date}T00:00:00`
      payload.rfp_submission_date = new Date(dateStr).toISOString()
    }
    if (data.next_task_due !== undefined && data.next_task_due !== null && data.next_task_due !== '') {
      const dateStr = data.next_task_due.includes('T') ? data.next_task_due : `${data.next_task_due}T00:00:00`
      payload.next_task_due = new Date(dateStr).toISOString()
    }
    if (data.award_date !== undefined && data.award_date !== null && data.award_date !== '') {
      const dateStr = data.award_date.includes('T') ? data.award_date : `${data.award_date}T00:00:00`
      payload.award_date = new Date(dateStr).toISOString()
    }
    
    logApiCall('PUT', `/opportunities/opportunities/${id}`, payload)
    
    try {
      const response = await api.put<Opportunity>(`/opportunities/opportunities/${id}`, payload)
      logApiResponse(response)
      return response.data
    } catch (error) {
      logApiError(error)
      throw error
    }
  },

  getTimeline: async (id: string): Promise<{ timeline: any[] }> => {
    const response = await api.get<{ timeline: any[] }>(`/opportunities/opportunities/${id}/timeline`)
    return response.data
  },

  delete: async (id: string): Promise<{ message: string; id: string }> => {
    const response = await api.delete<{ message: string; id: string }>(`/opportunities/opportunities/${id}`)
    return response.data
  },
}

